### ISSUE TITLE
How to instantiate multiple apollo clients for one app?

### ISSUE THREAD

**Issue by mlukaszczyk** [apollostack/apollo#60]
Hey all,

in our web app we need to use multiple graphql servers. Is there a documented way on doing this? Is it allowed to have multiple "nested" apollo providers without overriding the wrapping provider/s? 

Thanks,

Michael

**Comment by stubailo** [comment]
Right now you can only have one client in a certain React subtree when using `react-apollo`. Curious, why use multiple graphql servers? Seems like you lose a lot of the benefits of GraphQL when using it in that way.

Do you need to call queries on multiple servers in the same component? Just curious to get more information about this use case.

**Comment by mlukaszczyk** [comment]
Dear Sashko,

in our WebApp you have the ability to generate a postgreSQL db schema which is then stored in a mongoDB. This schema is then used with postgraphQL to generate a new graphQL server / endpoint. So we have a graphQL server for managing the schemas and always a "fresh" generated graphQL server for CRUD data handling on the postgres side, as soon as the schema service invokes a change to postgraphQL. 

Then we want to build a react component, which is suppossed to build a dynamic-data-table-view that is fetching the schema information from the mongoDB server to build the table frame (column and header information) and the data rows from the postgres server.

In theory, we don´t need both of the data in one component. We will try a different provider nesting.

best regards,

Michael Lukaszczyk

**Comment by jbaxleyiii** [comment]
@mlukaszczyk since react Apollo stores the client as `clent` in the context, nested providers will have collision so this won't be possible as it's currently built

**Comment by dallonf** [comment]
Maybe a better solution is to support easily composing two or more GraphQL endpoints into one? That way you can use GraphQL all the way down through various abstractions. I think there's a ticket somewhere to that effect, don't have time to search for it just now though.

**Comment by mlukaszczyk** [comment]
The solution of having the clients seperated in different subtrees worked for us, thanks! 

But still, endpoint composition would be a nice feature! (:

**Comment by stubailo** [comment]
Yeah I think it would be great to have a generic server that you can point at two GraphQL endpoints and it would just proxy the queries and have one schema that is the union of the two.

**Comment by austenmc** [comment]
There are other legitimate uses for supporting multiple schemas/clients. In the game I'm building w/ some friends, admins can set up their own server and the mobile client connects to whichever server the user provides. In this case, they have the same schema but are distinct servers, so caching needs to be kept separate.

Planning to solve this currently by doing the limited multi-server queries by hand and dropping into an Apollo client once the user select a server on which to play. Unclear whether I'll have to clear the caches every time a user switches games.

**Comment by n2liquid** [comment]
One word, lots of hype, pain, and accidental complexity: Microservices.

Is Apollo GraphQL not compatible with GraphQL microservices? What if I need to access data from multiple microservices in the same subtree?

**Comment by pedrospecter** [comment]
better word…dotnet!

leave the Apollo and JS. Learn dotnet and "link composition" is not so difficult afterall
